doc/man/man1/mapitest.1
.\" OpenChange Project Tools Man Pages
.\"
.\" This manpage is Copyright (C) 2008, 2010 Brad Hards
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\" 
.\" Since the OpenChange and Samba4 libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" Process this file with
.\" groff -man -Tascii mapitest.1
.\"
.TH MAPITEST 1 2013-01-24 "OpenChange 2.0 QUADRANT" "OpenChange Users' Manual"

.SH NAME
mapitest \- OpenChange torture test utility

.SH SYNOPSIS
.nf
mapitest [-?|--help] [--usage] [-f|--database=STRING] [-p|--profile=STRING]
  [-p|--password=STRING] [--confidential] [--color] [--subunit]
  [-o|--outfile=STRING] [--mapi-calls=STRING] [--list-all] [--no-server]
  [--dump-data] [-d|--debuglevel=STRING]
.fi

.SH DESCRIPTION
mapitest is a test harness / utility used for verifying correct operation
of various ExchangeRPC calls / MAPI functions provided by the OpenChange
MAPI libraries.

mapitest is not normally required by users, but you may be asked to provide
the output of mapitest for some kinds of bug investigations.

Note that mapitest performs a lot of transactions, including deleting folders
and messages. Unless you're very familiar with mapitest, we recommend only
using it with a test account.

.SH OPTIONS

.TP
.B --database
.TP
.B -f
Set the path to the profile database to use

.TP
.B --profile
.TP
.B -p
Set the profile to use. If no profile is specified, mapitest tries
to retrieve the default profile in the database. If no default profile
has been set, mapitest returns 
.B MAPI_E_NOT_FOUND .

.TP
.B --password
.TP
.B -P
Specify the password for the profile to use. This can be omitted if the
password is stored in the profile.

.TP
.B --confidential
Remove any sensitive data from the report.

.TP
.B --color
Use colors to indicate the results of each operation.

.TP
.B --subunit
Produce output in subunit protocol format, instead of the normal text
output. This disables colored output. Note that availability of this
option depends on suitable libraries being available at build time, so
check availability before relying on this.

.TP
.B --outfile
.TP
.B -o
Redirect the output of the tests to a file. The filename must be specified
as the argument to this option.

.TP
.B --mapi-calls
Run a specific test. The name of the test must be specified as the argument
to this option. See the --list-all option to obtain the name of the test.
This can be specified more than once in order to run a subset of tests.
This can also be used to run a "suite" of tests, by appending "-ALL" to
the name of the suite.

.TP
.B --list-all
Provide a list of all test suites and test names, along with a description
of the test. No tests will be run.

.TP
.B --no-server
Only run tests that do not require a server connection. This is the default
if a connection to the server cannot be established.

.TP
.B --dump-data
Dump the hex data. This is only required for debugging or educational purposes.

.TP
.B --debuglevel
.TP
.B -d
Set the debug level.

.SH EXAMPLES

.B Run all tests
.nf
mapitest
.fi

.B Only run two specific tests
.nf
mapitest --mapi-calls=NOSERVER-SROWSET --mapi-calls=OXCPRPT-GET-PROPS
.fi

.B Run all the NSPI tests
.nf
mapitest --mapi-calls=NSPI-ALL
.fi

.SH REMARKS
If you are using the default profile database path and have set a
default profile (using
.B mapiprofile --profile=profile_name -S
) you do not need to specify these parameters on the command line.

.SH AUTHOR
mapitest was written by Julien Kerihuel <j.kerihuel at openchange dot org>
with contributions from other OpenChange developers.

This man page was written by Brad Hards <bradh at openchange dot org>


ChangeLog
2010-08-29
    jkerihuel
	[r2144]
		Add CopyProfile and DuplicateProfile to IProfAdmin interface

2010-08-27
    jkerihuel
	[r2142]
		Add the localaddress patch prior Samba4 integration. Pending new
		Samba4 localaddress patch fixing memory leaks in the pipe.
	[r2141]
		Include a mapitest test suite / challenge for lzxpress
		
		The idea of this test is to check if lzxpress compressed blobs
		provided by OpenChange are the same than Outlook ones. The test takes
		a Outlook->Exchange capture file with compressed blob, pull the
		content, decompressed the content, recompress the content and check
		the output for differing bytes.
		
		This challenge code should give us good information when lzxpress
		compress code is fixed. Additionally you can uncomment the compression
		code in libmapi/emsmdb.c to do further testing.
	[r2140]
		- Add implementation for EcDoRpcExt2 (0xB) RPC opnum. Current code
		  doesn't provide compression for request and use obfuscation
		  instead. However mapi_response makes use of xpress and are
		  decompressed properly.
		
		- EcDoRpcExt2 implemented without subcontexts but uint8_t array
		
		- add a wrapper for emsmdb transaction to switch between
		  EcDoConnect/EcDoRpc and EcDoConnectEx/EcDoRpcExt2 depending on MAPI
		  profile 'exchange version' attribute.
		
		NOTE: emsmdb.c implements the logic for using compressed blob and
		checks whether compressed or obfuscation should be used depending on
		blob length. However this behavior has been commented out because
		comrpessed blobs are rejected by Exchange (ecRpcFormat 0x4B6 error).
2010-08-26
    jkerihuel
	[r2139]
		Reference profile structure in mapitest structure.
		
		Use profile->mailbox instead of szDisplayName for sending mapitest
		emails. This also fix ResolveNames mapitest issues when conflicts
		between usernames exist (e.g. user1, user11, user100).
    bradh
	[r2137]
		Prevent segfault if we don't have any appointments.
		
		Resolves issue #284.
		
		Also prevent leaking talloc context in a couple of error cases.
	[r2136]
		Merge Rev 2133 from branches/current-samba4 to trunk.
		
		I inadvertently committed this to the wrong branch.
		
		This pushes the fix for #229 into trunk.

2010-08-23
    jelmer
	[r2134]
		Update bzr ignores.

2010-08-18
    jkerihuel
	[r2131]
		Add a git update case in installsamba.sh script and make it available
		within OpenChange Makefile.

2010-08-17
    jkerihuel
	[r2127]
		Manually commit remaining changes from current-samba4 into trunk
	[r2126]
		Merge current-samba4 branch back into trunk

2010-08-11
    bradh
	[r2117]
		Microsoft confirmed that RopGetValidAttachments isn't supported in Exchange 2010.
	[r2114]
		fix missing part of r2113.
	[r2113]
		Fixes to better work with Exchange 2010.
		
		Still some open issues to resolving bug 227.
	[r2112]
		Fix handling of "not applicable" tests when using --mapi-calls
		
		Prior to this, if you asked for a test that was after a test that was
		flagged as suppressed, the test you asked for didn't get run, and the
		flagged test was reported instead.
2010-08-10
    bradh
	[r2111]
		Correctly handle cases where the start date of the range is after the first calendar appointment.
		
		Resolves bug:272.
		
		Thanks to Eric Reischer for reporting the problem and testing the fix.

2010-08-03
    bradh
	[r2107]
		trivial comment cleanup.

2010-08-01
    jkerihuel
	[r2105]
		Move from auto-generated mapicode.h to static code.
	[r2104]
		Add configure.ac support for gthread implementation but give preferences to pthread
		Add THREAD_CFLAGS to config.mk.in
		ifdef between pthread and gthread in libmapi_private.h
	[r2103]
		Check for libpthread in configure.ac and make this lib optional
		Add references to config.h in libmapi_private.h
		Replace references to config.h with libmapi_private.h

2010-07-31
    jkerihuel
	[r2101]
		- Add an exchange-version parameter to mapiprofile structure and tool
		  which let user choose between 0x0/0x2 or 0xA/0xB
		- Tweak libmapi implementation to allow EcDoConnectEx usage
		- Improve mapiprofile --dump to show more useful parameters
	[r2100]
		- Add custom NDR implementation for EcDoConnectEx to bypass pointers/array problems
		- make use of mapi2k7_AuxInfo structure for rgbAuxOut
		
		- Create a subcontext for AUX_HEADER structures to jump over unknown cases properly.
	[r2099]
		Handle the no obfuscation nor compression RPC_HEADER_EXT flags case 
    bradh
	[r2098]
		Trivial api documentation change.

2010-07-29
    bradh
	[r2095]
		Another trivial API documentation fix.
	[r2094]
		More trivial API documentation fixes.
	[r2093]
		Minor API documentation fix.
	[r2092]
		Implement some null pointer checks.
		
		Also rewords some API documentation.
		
		Original issue and patch by Erik Hovland - thanks again. I implemented it
		with some slight changes.

2010-07-28
    bradh
	[r2090]
		Fix RecipientRow IDL to match MS-OXCDATA 2.9.3
		
		This allows handling of addresses that contain X400 format content.
		
		Patch by Samuel Sieb and I, resolves ticket #250.

2010-07-26
    jkerihuel
	[r2087]
		Fix a crash in openchangeclient vs Exchange 2010:
		When Queryrows is called on Exchange 2010 and no rows are available, QueryRows returns MAPI_E_NOT_IMPLEMENTED error (0x8004FFFF).
		Ensure we have at least one message to fetch in openchangeclient --fetchmail prior calling QueryRows.

2010-07-24
    jkerihuel
	[r2079]
		Very basic use of OpenUserMailbox in openchangeclient.
		Doesn't handle clustered Exchange servers.
    bradh
	[r2077]
		Rename a few function arguments to avoid doxygen warnings.
		
		This is a workaround, but will allow easier checking for the future.
	[r2076]
		Update doxygen configuration to match latest version
	[r2075]
		Add missing doxygen parameter information.
	[r2074]
		Fix a couple of doxygen issues.

2010-07-23
    jkerihuel
	[r2072]
		Use sed in Makefile to replace local headers reference in installed headers
	[r2071]
		Use local headers rather than installed ones

utils/mapitest/mapitest.c
/*
   Stand-alone MAPI testsuite

   OpenChange Project

   Copyright (C) Julien Kerihuel 2008

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "utils/mapitest/mapitest.h"
#include "utils/openchange-tools.h"

#include <popt.h>
#include <param.h>

#include "config.h"

/**
	\file
	Core of %mapitest implementation
*/

/**
   Initialize %mapitest structure
 */
static void mapitest_init(TALLOC_CTX *mem_ctx, struct mapitest *mt)
{
	mt->mem_ctx = mem_ctx;
	mt->stream = NULL;
	memset(&mt->info, 0, sizeof (mt->info));
	mt->session = NULL;

	mt->session = NULL;
	mt->mapi_all = true;
	mt->confidential = false;
	mt->no_server = false;
	mt->color = false;
	mt->online = false;
	mt->mapi_suite = false;
	mt->cmdline_calls = NULL;
	mt->cmdline_suite = NULL;
	mt->subunit_output = false;
}

/**
  Initialize %mapitest output stream

  \param mt pointer to mapitest context
  \param filename filename to write to (can be null, for output to stdout)
*/
static void mapitest_init_stream(struct mapitest *mt, const char *filename)
{
	if (filename == NULL) {
		mt->stream = fdopen(STDOUT_FILENO, "a");
	} else {
		mt->stream = fopen(filename, "w+");
	}

	if (mt->stream == NULL) {
		err(errno, "fdopen/fopen");
	}
}

/**
  Clean up %mapitest output stream

  \param mt pointer to mapitest context
*/
static void mapitest_cleanup_stream(struct mapitest *mt)
{
	fclose(mt->stream);
}


static bool mapitest_get_testnames(TALLOC_CTX *mem_ctx, struct mapitest *mt,
				   const char *parameter)
{
	struct mapitest_unit	*el = NULL;
	char			*temptok = NULL;

	if ((temptok = strtok((char *)parameter, ";")) == NULL) {
		fprintf(stderr, "Invalid testname list [;]\n");
		return false;
	}

	el = talloc_zero(mem_ctx, struct mapitest_unit);
	el->name = talloc_strdup(mem_ctx, temptok);
	DLIST_ADD(mt->cmdline_calls, el);

	while ((temptok = strtok(NULL, ";")) != NULL) {
		el = talloc_zero(mem_ctx, struct mapitest_unit);
		el->name = talloc_strdup(mem_ctx, temptok);
		DLIST_ADD_END(mt->cmdline_calls, el, struct mapitest_unit *);
	}

	return true;
}


static void mapitest_list(struct mapitest *mt, const char *name)
{
	struct mapitest_suite		*sel;
	struct mapitest_test		*el;

	/* List all tests */
	if (!name) {
		for (sel = mt->mapi_suite; sel; sel = sel->next) {
			printf("[*] Suite %s\n", sel->name);
			printf("===================================\n");
			printf("    * %-15s %s\n", "Name:", sel->name);
			printf("    * %-15s %5s\n", "Description:", sel->description);
			printf("    * Running Tests:\n");
			for (el = sel->tests; el; el = el->next) {
				printf("\t    - %-35s: %-10s\n", el->name, el->description);
			}
			printf("\n\n");
		}
	}
}


/**
 * Retrieve server specific information
 */
static bool mapitest_get_server_info(struct mapitest *mt,
				     char *opt_profname,
				     const char *password,
				     bool opt_dumpdata,
				     const char *opt_debug)
{
	TALLOC_CTX		*mem_ctx;
	enum MAPISTATUS		retval;
	struct emsmdb_info	*info = NULL;
	struct mapi_session	*session = NULL;

	/* if the user explicitly asked for just the no-server tests 
	to be run, then we're done here */
	if (mt->no_server == true) return 0;

	mem_ctx = talloc_named(NULL, 0, "mapitest_get_server_info");

	/* if no profile was specified, get the default */
	if (!opt_profname) {
		retval = GetDefaultProfile(mt->mapi_ctx, &opt_profname);
		if (retval != MAPI_E_SUCCESS) {
			mapi_errstr("GetDefaultProfile", retval);
			talloc_free(mem_ctx);
			return false;
		}
	}
		

	/* debug options */
	SetMAPIDumpData(mt->mapi_ctx, opt_dumpdata);

	if (opt_debug) {
		SetMAPIDebugLevel(mt->mapi_ctx, atoi(opt_debug));
	}

	retval = MapiLogonEx(mt->mapi_ctx, &session, opt_profname, password);
	MAPIFreeBuffer(opt_profname);
	talloc_free(mem_ctx);
	if (retval != MAPI_E_SUCCESS) {
		mapi_errstr("MapiLogonEx", retval);
		return false;
	}
	mt->session = session;
	mt->profile = session->profile;

	info = emsmdb_get_info(session);
	memcpy(&mt->info, info, sizeof (struct emsmdb_info));

	/* extract org and org_unit from info.mailbox */
	mt->org = x500_get_dn_element(mt->mem_ctx, info->szDNPrefix, "/o=");
	mt->org_unit = x500_get_dn_element(mt->mem_ctx, info->szDNPrefix, "/ou=");
	
	return true;
}



/**
 *  main program
 */
int main(int argc, const char *argv[])
{
	enum MAPISTATUS		retval;
	int32_t			num_tests_failed;
	TALLOC_CTX		*mem_ctx;
	struct mapitest		mt;
	poptContext		pc;
	int			opt;
	bool			ret;
	bool			opt_dumpdata = false;
	const char     		*opt_debug = NULL;
	const char		*opt_profdb = NULL;
	char			*opt_profname = NULL;
	const char		*opt_password = NULL;
	const char		*opt_outfile = NULL;
	char			*prof_tmp = NULL;
	bool			opt_leak_report = false;
	bool			opt_leak_report_full = false;

	enum { OPT_PROFILE_DB=1000, OPT_PROFILE, OPT_PASSWORD,
	       OPT_CONFIDENTIAL, OPT_OUTFILE, OPT_MAPI_CALLS,
	       OPT_NO_SERVER, OPT_LIST_ALL, OPT_DUMP_DATA,
	       OPT_DEBUG, OPT_COLOR, OPT_SUBUNIT, OPT_LEAK_REPORT,
	       OPT_LEAK_REPORT_FULL };

	struct poptOption long_options[] = {
		POPT_AUTOHELP
		{ "database",        'f', POPT_ARG_STRING, NULL, OPT_PROFILE_DB,       "set the profile database", NULL },
		{ "profile",         'p', POPT_ARG_STRING, NULL, OPT_PROFILE,          "set the profile name", NULL },
		{ "password",        'p', POPT_ARG_STRING, NULL, OPT_PASSWORD,         "set the profile or account password", NULL },
		{ "confidential",      0, POPT_ARG_NONE,   NULL, OPT_CONFIDENTIAL,     "remove any sensitive data from the report", NULL },
		{ "color",             0, POPT_ARG_NONE,   NULL, OPT_COLOR,            "color MAPI retval", NULL },
#if HAVE_SUBUNIT
		{ "subunit",           0, POPT_ARG_NONE,   NULL, OPT_SUBUNIT,          "output in subunit protocol format", NULL },
#endif
		{ "outfile",         'o', POPT_ARG_STRING, NULL, OPT_OUTFILE,          "set the report output file", NULL },
		{ "mapi-calls",        0, POPT_ARG_STRING, NULL, OPT_MAPI_CALLS,       "test custom ExchangeRPC tests", NULL },
		{ "list-all",          0, POPT_ARG_NONE,   NULL, OPT_LIST_ALL,         "list suite and tests - names and description", NULL },
		{ "no-server",         0, POPT_ARG_NONE,   NULL, OPT_NO_SERVER,        "only run tests that do not require server connection", NULL },
		{ "dump-data",         0, POPT_ARG_NONE,   NULL, OPT_DUMP_DATA,        "dump the hex data", NULL },
		{ "debuglevel",      'd', POPT_ARG_STRING, NULL, OPT_DEBUG,            "set debug level", NULL },
		{ "leak-report",       0, POPT_ARG_NONE,   NULL, OPT_LEAK_REPORT,      "enable talloc leak reporting on exit", NULL },
		{ "leak-report-full",  0, POPT_ARG_NONE,   NULL, OPT_LEAK_REPORT_FULL, "enable full talloc leak reporting on exit", NULL },
		POPT_OPENCHANGE_VERSION
		{ NULL, 0, 0, NULL, 0, NULL, NULL }
	};

	mem_ctx = talloc_named(NULL, 0, "mapitest");
	mapitest_init(mem_ctx, &mt);
	mapitest_register_modules(&mt);

	pc = poptGetContext("mapitest", argc, argv, long_options, 0);

	while ((opt = poptGetNextOpt(pc)) != -1) {
		switch (opt) {
		case OPT_DUMP_DATA:
			opt_dumpdata = true;
			break;
		case OPT_DEBUG:
			opt_debug = poptGetOptArg(pc);
			break;
		case OPT_PROFILE_DB:
			opt_profdb = poptGetOptArg(pc);
			break;
		case OPT_PROFILE:
			prof_tmp = poptGetOptArg(pc);
			opt_profname = talloc_strdup(mem_ctx, prof_tmp);
			free(prof_tmp);
			prof_tmp = NULL;
			break;
		case OPT_PASSWORD:
			opt_password = poptGetOptArg(pc);
			break;
		case OPT_CONFIDENTIAL:
			mt.confidential = true;
			break;
		case OPT_OUTFILE:
			opt_outfile = poptGetOptArg(pc);
			break;
		case OPT_MAPI_CALLS:
			ret = mapitest_get_testnames(mem_ctx, &mt, poptGetOptArg(pc));
			if (ret == false) exit (-1);
			mt.mapi_all = false;
			break;
		case OPT_NO_SERVER:
			mt.no_server = true;
			break;
		case OPT_COLOR:
			mt.color = true;
			break;
		case OPT_SUBUNIT:
			mt.subunit_output = true;
			break;
		case OPT_LIST_ALL:
			mapitest_list(&mt, NULL);
			talloc_free(mem_ctx);
			poptFreeContext(pc);
			return 0;
			break;
		case OPT_LEAK_REPORT:
			opt_leak_report = true;
			talloc_enable_leak_report();
			break;
		case OPT_LEAK_REPORT_FULL:
			opt_leak_report_full = true;
			talloc_enable_leak_report_full();
			break;
		}
	}

	poptFreeContext(pc);

	/* Sanity check */
	if (mt.cmdline_calls && (mt.mapi_all == true)) {
		fprintf(stderr, "mapi-calls and mapi-all can't be set at the same time\n");
		return -1;
	}

	/* Initialize MAPI subsystem */
	if (!opt_profdb) {
		opt_profdb = talloc_asprintf(mem_ctx, DEFAULT_PROFDB, getenv("HOME"));
	}

	retval = MAPIInitialize(&(mt.mapi_ctx), opt_profdb);
	if (retval != MAPI_E_SUCCESS) {
		mapi_errstr("MAPIInitialize", retval);
		return -2;
	}

	mapitest_init_stream(&mt, opt_outfile);
	
	mt.online = mapitest_get_server_info(&mt, opt_profname, opt_password,
					     opt_dumpdata, opt_debug);

	mapitest_print_headers(&mt);

	/* Do not run any tests if we couldn't find a profile or if
	 * server is offline and connection to server was implicitly
	 * specified */
	if (!opt_profname && mt.online == false && mt.no_server == false) {
		fprintf(stderr, "No MAPI profile found for online tests\n");
		return -2;
	}

	/* Run custom tests */
	if (mt.cmdline_calls) {
		struct mapitest_unit	*el;
		
		for (el = mt.cmdline_calls; el; el = el->next) {
			printf("[*] %s\n", el->name);
			mapitest_run_test(&mt, el->name);
		}
	} else {
		mapitest_run_all(&mt);
	}

	num_tests_failed = mapitest_stat_dump(&mt);

	mapitest_cleanup_stream(&mt);

	/* Uninitialize and free memory */
	MAPIUninitialize(mt.mapi_ctx);
	talloc_free(mt.mem_ctx);

	if (opt_leak_report) {
		talloc_report(NULL, stdout);
	}

	if (opt_leak_report_full) {
		talloc_report_full(NULL, stdout);
	}

	return num_tests_failed;
}
